### PR TITLE
More consistent parts removal from memory and from disk

### PR DESCRIPTION
**Pull request by alesapin**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

### CODE REVIEW

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 1]
```diff
@@ -2153,10 +2148,33 @@ void MergeTreeData::dropAllData()
...
+        /// Parts removal process can be important and on the next try it's better to try to remove
+        /// them instead of remove recursive call.
+        LOG_WARNING(log, "dropAllData: got exception removing parts from disk, removing successfully removed parts from memory.");
+        for (const auto & part : all_parts)
+        {
+            if (!part_names_failed.contains(part->name))
+                data_parts_indexes.erase(part->info);
+        }
+
+        throw;
```
Comment (tavplubix) [inline src/Storages/MergeTree/MergeTreeData.cpp:2174]: FYI it may cause some issues with Ordinary database (but I don't care)

Comment (tavplubix) [reply src/Storages/MergeTree/MergeTreeData.cpp:2174]: Do we still need `removeRecursive` below?

**Review by tavplubix — APPROVED** [review]

File: src/Storages/MergeTree/MergeTreeData.cpp (hunk 1 above)
Comment (alesapin) [reply src/Storages/MergeTree/MergeTreeData.cpp:2174]: I don't know, only as safe-guard against some garbage like /detached?

**Comment by alesapin** [comment]
`ClickHouse Stress Test (ubsan)`
Hunged query: `DROP DATABASE test_lj946rac`
gdb traces: https://pastila.nl/?0006a09d/6af108cc35ab23de23d70ffac7d07304
logs:
```
2022.09.09 21:54:05.637094 [ 1657 ] {} <Trace> test_lj946rac.alter_table_2 (12405db5-f341-4863-887f-777e29731a68): dropAllData: waiting for locks.
2022.09.09 21:54:05.637120 [ 1657 ] {} <Trace> test_lj946rac.alter_table_2 (12405db5-f341-4863-887f-777e29731a68): dropAllData: removing data parts (count 37) from filesystem.
2022.09.09 21:54:05.637131 [ 1657 ] {} <Debug> test_lj946rac.alter_table_2 (12405db5-f341-4863-887f-777e29731a68): Removing part from filesystem -1_0_0_0_3
2022.09.09 21:54:05.637191 [ 1657 ] {} <Trace> test_lj946rac.alter_table_2 (12405db5-f341-4863-887f-777e29731a68): Part -1_0_0_0_3 local references is zero, will check blobs can be removed in zookeeper
2022.09.09 21:54:05.637433 [ 1657 ] {} <Warning> test_lj946rac.alter_table_2 (12405db5-f341-4863-887f-777e29731a68): dropAllData: got exception removing parts from disk, removing successfully removed parts from memory.
2022.09.09 21:54:05.637558 [ 1657 ] {} <Error> DatabaseCatalog: Cannot drop table test_lj946rac.alter_table_2 (12405db5-f341-4863-887f-777e29731a68). Will retry later.: Code: 999. Coordination::Exception: Session expired, path: Session expired. (KEEPER_EXCEPTION), Stack trace (when copying this message, always include the lines below):
```

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 2]
```diff
@@ -822,6 +822,15 @@ void StorageReplicatedMergeTree::createReplica(const StorageMetadataPtr & metada
     } while (code == Coordination::Error::ZBADVERSION);
 }
 
+
+zkutil::ZooKeeperPtr StorageReplicatedMergeTree::getZooKeeperIfTableShutDown() const
+{
+    if (zookeeper_name == default_zookeeper_name)
+        return getContext()->getZooKeeper();
+    else
+        return getContext()->getAuxiliaryZooKeeper(zookeeper_name);
```
Comment (tavplubix) [inline src/Storages/StorageReplicatedMergeTree.cpp:831]:
```suggestion
    zkutil::ZooKeeperPtr maybe_new_zookeeper;
    if (zookeeper_name == default_zookeeper_name)
        maybe_new_zookeeper = getContext()->getZooKeeper();
    else
        maybe_new_zookeeper = getContext()->getAuxiliaryZooKeeper(zookeeper_name);
    maybe_new_zookeeper->sync(zookeeper_path);
    return maybe_new_zookeeper;
```

**Comment by tavplubix** [comment]
Tests on the prev commit were ok; Fast Test passed on the last commit, let's merge

**Pull request merged**